### PR TITLE
SLI-37 Generated sources take too long to analyse

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 1]
```diff
@@ -136,31 +138,44 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
       SourceFolder[] sourceFolders = e.getSourceFolders();
       for (SourceFolder sourceFolder : sourceFolders) {
-        if (sourceFolder.getFile() == null) {
+        if (sourceFolder.getFile() == null || sourceFolder.isSynthetic()) {
           continue;
         }
+
         if (VfsUtil.isAncestor(sourceFolder.getFile(), file, false)) {
+          if (isExcludedOrUnderExcludedDirectory(file, e)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing excluded file: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 2]
```diff
@@ -136,31 +138,44 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
-    if ("java".equalsIgnoreCase(file.getFileType().getDefaultExtension()) && !isSource(file, module)) {
-      SonarLintConsole.get(module.getProject()).debug("Not automatically analysing java file outside source folder: " + file.getName());
-      return false;
-    }
-
-    return true;
-  }
-
-  public static boolean isSource(VirtualFile file, Module module) {
     ModuleRootManager moduleRootManager = ModuleRootManager.getInstance(module);
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: CRITICAL Parameter 1 to this call is marked "org.jetbrains.annotations.NotNull" but null is passed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2637)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 3]
```diff
@@ -136,31 +138,44 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+        if (sourceFolder.getFile() == null || sourceFolder.isSynthetic()) {
           continue;
         }
+
         if (VfsUtil.isAncestor(sourceFolder.getFile(), file, false)) {
+          if (isExcludedOrUnderExcludedDirectory(file, e)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing excluded file: " + file.getName());
+            return false;
+          } else if (isJavaResource(sourceFolder)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing file under resources: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
BLOCKER NullPointerException might be thrown as 'module' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 4]
```diff
@@ -136,31 +138,44 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+            return false;
+          } else if (isJavaGeneratedSource(sourceFolder)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing file belonging to generated source folder: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'module' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 5]
```diff
@@ -136,31 +138,44 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+
     for (ContentEntry e : entries) {
       SourceFolder[] sourceFolders = e.getSourceFolders();
       for (SourceFolder sourceFolder : sourceFolders) {
-        if (sourceFolder.getFile() == null) {
+        if (sourceFolder.getFile() == null || sourceFolder.isSynthetic()) {
           continue;
         }
+
         if (VfsUtil.isAncestor(sourceFolder.getFile(), file, false)) {
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: CRITICAL Parameter 2 to this call is marked "org.jetbrains.annotations.NotNull" but null is passed. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2637)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* BLOCKER 5 blocker
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/eb069decbd709030a6741a363f6d972cadc9288f/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/eb069decbd709030a6741a363f6d972cadc9288f/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): The Cyclomatic Complexity of this method "shouldAnalyzeAutomatically" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 6]
```diff
@@ -136,31 +138,46 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+          if (isJavaResource(sourceFolder)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing file under resources: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
BLOCKER NullPointerException might be thrown as 'module' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 7]
```diff
@@ -136,31 +138,46 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+            return false;
+          } else if (isJavaGeneratedSource(sourceFolder)) {
+            SonarLintConsole.get(module.getProject()).debug("Not automatically analysing file belonging to generated source folder: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'module' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 8]
```diff
@@ -136,31 +138,46 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
-    return true;
-  }
-
-  public static boolean isSource(VirtualFile file, Module module) {
     ModuleRootManager moduleRootManager = ModuleRootManager.getInstance(module);
     ContentEntry[] entries = moduleRootManager.getContentEntries();
+
     for (ContentEntry e : entries) {
+      if (isExcludedOrUnderExcludedDirectory(file, e)) {
+        SonarLintConsole.get(module.getProject()).debug("Not automatically analysing excluded file: " + file.getName());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'module' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* BLOCKER 5 blocker
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/bcfd1a5349ca1d9d105872f792d30f2a8c793485/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/bcfd1a5349ca1d9d105872f792d30f2a8c793485/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): The Cyclomatic Complexity of this method "shouldAnalyzeAutomatically" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 9]
```diff
@@ -46,6 +46,8 @@
 import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.jps.model.java.JavaModuleSourceRootTypes;
```
Comment (henryju) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: Will it work on all Jetbrains IDEs (PHPStorm, Webstorm, ... ?)

**Comment by henryju** [comment]
LGTM if you are sure it works on all Jetbrains IDEs

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java [hunk 10]
```diff
@@ -136,31 +138,46 @@ public static boolean shouldAnalyzeAutomatically(@Nullable VirtualFile file, @Nu
...
+          }
+
           return true;
         }
       }
     }
 
-    return false;
+    // java must be in a source root. For other files, we always analyse them.
+    return !"java".equalsIgnoreCase(file.getFileType().getDefaultExtension());
```
Comment (SonarTech) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: BLOCKER NullPointerException might be thrown as 'file' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* BLOCKER 6 blocker
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/546c45a3725ee893c2f11f31dc6be099f7d761f2/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1142)
1. MAJOR [SonarLintUtils.java#L135](https://github.com/SonarSource/sonarlint-intellij/blob/546c45a3725ee893c2f11f31dc6be099f7d761f2/src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java#L135): The Cyclomatic Complexity of this method "shouldAnalyzeAutomatically" is 15 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java (hunk 9 above)
Comment (dbmeneses) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: Good catch. It does not work on PHPStorm. And the worst is that there is no runtime error/logging happening, it simply does not work.

Comment (dbmeneses) [inline src/main/java/org/sonarlint/intellij/util/SonarLintUtils.java]: Nevermind, it works, I didn't test it properly.

**Pull request merged**